Homework2/Task01/test01.py

```diff
@@ -0,0 +1,120 @@
+import os
+
+from Task01.task01 import (
+    count_non_ascii_chars,
+    count_punctuation_chars,
+    get_longest_diverse_words,
+    get_most_common_non_ascii_char,
+    get_rarest_char,
+    open_file,
+)
+
+import pytest
+
+
+@pytest.fixture()
+def file_test_open():
+    filename = "file_test_open.txt"
+    with open(filename, "w+") as f:
+        f.write(
+            "Der Waldgang \\u2014 es ist keine Idylle, die sich hinter dem Titel verbirgt. "
+        )
+    yield filename
+    os.remove(filename)
+
+
+def test_open_file(file_test_open):
+    actual_result = open_file(file_test_open)
+
+    assert (
+        actual_result
+        == "Der Waldgang \u2014 es ist keine Idylle, die sich hinter dem Titel verbirgt. "
+    )
+
+
+@pytest.fixture()
+def file_with_diverse_words():
+    filename = "file_with_diverse_words.txt"
+    with open(filename, "w+") as f:
+        f.write(
+            "F\\u00e4lle abcd abcde aaaaaaaaaaaaaaaaaaaa abcdef abcdefg"
+            " abcdefgh abcdefghi  abcdefghij abcdefghijk abcdefghijkl"
+        )
+    yield filename
+    os.remove(filename)
+
+
+def test_get_longest_diverse_words(file_with_diverse_words):
+    actual_result = get_longest_diverse_words(file_with_diverse_words)
+
+    assert actual_result == [
+        "abcdefghijkl",
+        "abcdefghijk",
+        "abcdefghij",
+        "abcdefghi",
+        "abcdefgh",
+        "abcdefg",
+        "abcdef",
+        "abcde",
+        "F\u00e4lle",
+        "abcd",
+    ]
+
+
+@pytest.fixture()
+def file_with_rarest_char():
+    filename = "file_with_rarest_char.txt"
+    with open(filename, "w+") as f:
+        f.write("madam I'm Adam")
+    yield filename
+    os.remove(filename)
+
+
+def test_get_rarest_char(file_with_rarest_char):
+    actual_result = get_rarest_char(file_with_rarest_char)
+
+    assert actual_result == "I"
+
+
+@pytest.fixture()
+def file_with_punctuation():
+    filename = "file_with_rarest_char.txt"
+    with open(filename, "w+") as f:
```

Comment:
What is `w+` ?

Reply:
Opens a file for both writing and reading. Overwrites the existing file if the file exists. If the file does not exist, creates a new file for reading and writing.

Reply:
Why do you need it open for reading?  Do you read it before closing?

Reply:
No, I don't. I will fix it.